### ISSUE TITLE
EventSourcedBehavior.lastSequenceNumber incorrect values after recovery

### ISSUE THREAD

**Issue by aludwiko** [akka/akka#32359]
Calling `EventSourcedBehavior.lastSequenceNumber` in the event handler returns incorrect value after entity recovery:
Example for a new entity:
```
cmd handler: 0
evt handler: 0

cmd handler: 1
evt handler: 2

cmd handler: 2
evt handler: 3
```
After recovery:
```
evt handler: 1
evt handler: 2
evt handler: 3

cmd handler: 3
evt handler: 3

cmd handler: 4
evt handler: 5
```

When debugging, I've noticed that instead of the first event handler (after recovery) will still use `akka.persistence.typed.internal.ReplayingEvents#currentSequenceNumber`, and the following one will switch to `akka.persistence.typed.internal.Running.HandlingCommands#currentSequenceNumber`.

**Comment by johanandren** [comment]
We have test coverage for sequence number on commands after recovery passing https://github.com/akka/akka/blob/main/akka-persistence-typed-tests/src/test/scala/akka/persistence/typed/scaladsl/EventSourcedSequenceNumberSpec.scala#L113 so there must be something more to what you observed. Was there anything special going on in the entity where you saw it @aludwiko ?

**Comment by patriknw** [comment]
I can confirm that I see this issues. For me it was lastSequenceNumber from event handler for the first event in replicated event sourcing, on the non-origin side. That matches above description that it happens after recovery.

**Comment by sebastian-alfers** [comment]
I have boiled down the example that Johan pointed to, to a very basic version, and the faulty behavior reported in this ticket starts to surface as soon as I remove the `receiveSignal`.

**Comment by sebastian-alfers** [comment]
This is the example I use: https://github.com/akka/akka/compare/main...sebastian-alfers:akka:wip-lastSequenceNumber-incorrect?expand=1

**Comment by sebastian-alfers** [comment]
It is not the `receiveSignal` that makes the difference, but the fact if I send a message (to the test probe) or not.

**Comment by johanandren** [comment]
Thanks to @sebastian-alfers digging here we have a more clear idea why/how this happens:

The test sends a command before the recovery is complete (race condition because depends on how fast recovery completes, but seems pretty consistent), since recovery is ongoing, the message is stashed.

When recovery completes the state the behavior the actual actor references, is still `ReplayingEvents`. When unstashing to handle the commands that was stashed during replay, it creates  `HandlingCommands`  a does a `tryUnstashOne` where it passes the `HandlingCommand` instance to, it has not yet actually switched behavior from the typed `ActorAdapter` point of view.

`lastSequenceNumber` is implemented by walking through the internals to find the current persistent actor state machine state, but since that was not updated, when handling the command, it still references the `ReplayingEvents` state and the sequence number from that (for empty journal 0), it does not increase when applying the event and this is where the incorrect number comes from.

Once that first event is handled, I _think_ it actually switches behavior from the adapter point of view, and that for a second stashed message, the sequence number in the event handler is correct.

I have no good idea of a way to solve this easily.

Not important, but for the record, the difference with sending a message to a probe or not was that the test would block until recovery completed instead of immediately sending a message so that it got stashed, therefore the probe would "fix" the race condition (or at least make it way less likely).

**Comment by sebastian-alfers** [comment]
Thanks for wring up this summary Johan. As discussed I captures my findings [in this draft-pr](https://github.com/akka/akka/pull/32651) (will close later) in case someone wants to pick up the investigation.

**Comment by sebastian-alfers** [comment]
The test in my closed draft-pr still fails, but with a different error. Basically, we do not have a test that does **not** run a `probe.expectMessage("0 onRecoveryComplete")`. 

Not doing this above check causes the actor to be have differently.

**Comment by patriknw** [comment]
Thanks, I'll try that

**Comment by sebastian-alfers** [comment]
You can reproduce if you clone https://github.com/sebastian-alfers/akka, branch `test-lastSequenceNumber-spec` and then:
```
sbt "akka-persistence-typed-tests/testOnly *.EventSourcedSequenceNumberBasicSpec"
```

**Comment by patriknw** [comment]
@sebastian-alfers That is because you are using the same persistenceId for the two tests. Then the second will start with replay of the event that was persisted by the previous test, and therefore event handler first.

Shouldn't the tests be:
```
    "be accessible accessible when testing RecoveryCompleted" in {
      runTheSpec(true, "1")
    }

    "be accessible accessible when NOT testing RecoveryCompleted" in {
      runTheSpec(false, "2")
    }
```

**Comment by sebastian-alfers** [comment]
You are right, it should be different (in fact, it was like this at some point, which is why I added the second parameter to `runTheSpec`).

But when I change it, it still fails. I pushed it change.

**Comment by sebastian-alfers** [comment]
Sorry for the trouble, I made a mistake when rebasing your fix from yesterday. Now it works.
